### ISSUE TITLE
Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by GrachevArtem** [zy4kamu/Coda#4]
Segfault during tokenization after calling Dictionary. 
If delete line with Dictionary everything is ok. 

```
# -*- coding: utf-8 -*-

from coda.tokenizer import Tokenizer
from coda.disambiguator import Disambiguator
from coda.dictionary import Dictionary


if __name__ == '__main__':
    tokenizer = Tokenizer("RU")
    disambiguator = Disambiguator("RU")
    wf_generator = Dictionary("RU")

    sentence = u'МИД пригрозил ограничить выезд американских дипломатов по России'
    tokens = tokenizer.tokenize(sentence)
```

**Comment by zy4kamu** [comment]
H,

I can take a look in the evening. Misha, I think Dictionary is your class, right?

**Comment by mkudinov** [comment]
@zy4kamu yes. it doesn't crash by itself. The problem emerges when both Tokenizer and Dictionary are referenced

**Comment by zy4kamu** [comment]
Wow, that is strange. Ok, I will take a look if you don´t fix it by the evening.

**Comment by zy4kamu** [comment]
Hi, guys. 

I haven't fixed the bug. Hope I will be able to do it tomorrow if one of you doesn't close it before. But I would like to tell something which may be helpful.

1. I have made debug and release version for python. Please, take a look on last version. Now if you run 'sudo bash build_cpp.sh -d' it will create all libs with gdb support in /opt/coda/debug. And then you will be able to debug c++ code from python extensions with any IDE you like (I still enjoy qt-creator). See bug_example.py.

2. I found that it falls on exit of Tokenizer/PythonExtension.cpp:Tokenize, i.e. it calculates correct output and then just exists with segmentation fault. Looks like UB (((

3. Then I commented lines 50-51 in DictionaryFacade.cpp and it works! On my side I create MorphologicalDictionary in DisambiguatorSettingsRus.cpp lines 70-71. It looks like memory leaks in Ha's dictionary, namely it can't be created twice.

In a nutshell, I think either we can debug Ha's dictionary which can be a real torture or try to make it singleton.

What do you think?

**Comment by mkudinov** [comment]
Well, I think there are two points: 
1)  I think, everyone agrees that UB is unacceptable. If we have it in basic modules we must fix it. Hopefully, Valgrind will help without deep debug.
2)  Loading memory consuming data structure twice is bad design. DictionaryFacade is rather clumsy by itself: actually it loads both MorphologicalDictionary and DictionaryInterface because of Ha's isoteric solutions. So storing MorphologicalDictionary also in Disambiguator and Parser is pure hell. We need to discuss how to resolve this issue.

I'll check DictionaryFacade with Valgrind

**Comment by mkudinov** [comment]
Okay, I guess the problem actually was with ctypes in Tokenizer class. The program crashes on tokenizer.py:71
content = ctypes.c_wchar_p(func(token_index)).value
Magnificent @ajkxyz proposed that the crash was caused by ctypes misbehaving. I replaced it with cffi based solution (like in Dictionary class). Now it works.
Also, with cffi c-code references in python become more compact and tidy. I believe we should completely drop ctypes. 
@zy4kamu Could you update Disambiguator module? 

So, the conclusion:
1. I propose complete switch to cffi based solution.
2. We should discuss problem of multiple dictionary copies in memory. Shall we start new issue?

**Comment by zy4kamu** [comment]
Hi,

1.  I propose complete switch to cffi based solution.

Agree. Create a new issue and I will switch to cfii based solution.

2. We should discuss problem of multiple dictionary copies in memory.

If you have time then it would be nice. If not, then "работает и ладно ..."

3. Could you clarify: is this a bug of ctypes? Yesterday in call stack I
saw that the problem came from cffi.  Is it possible that ctypes and cffi
are not compatible?

2017-02-02 15:51 GMT+01:00 mkudinov <notifications@github.com>:

**Comment by mkudinov** [comment]
2.  "работает и ладно ..."
3. Umm... I don't know :) But I still think that cffi is better

**Comment by GrachevArtem** [comment]
I get another error on this code

```
import coda.common

from coda.tokenizer import Tokenizer
from coda.disambiguator import Disambiguator
from coda.dictionary import Dictionary

if __name__ == '__main__':
    tokenizer = Tokenizer("RU")
    disambiguator = Disambiguator("RU")
    wf_generator = Dictionary("RU")

    sentence = u'МИД пригрозил ограничить выезд американских дипломатов по России'
    tokens = tokenizer.tokenize(sentence)
    disambiguated = disambiguator.disambiguate(tokens)
```

```
    disambiguated = disambiguator.disambiguate(tokens)
  File "build/bdist.linux-x86_64/egg/coda/disambiguator.py", line 85, in disambiguate
  File "build/bdist.linux-x86_64/egg/coda/tokenizer.py", line 35, in push_tokens_to_cpp
  File "/home/artem/anaconda3/envs/py27/lib/python2.7/site-packages/cffi/api.py", line 837, in __getattr__
    make_accessor(name)
  File "/home/artem/anaconda3/envs/py27/lib/python2.7/site-packages/cffi/api.py", line 832, in make_accessor
    raise AttributeError(name)
AttributeError: ResetParsedTokens

```

**Comment by mkudinov** [comment]
Mea culpa 
Already fixed

**Comment by zy4kamu** [comment]
Hi, 

Misha, did you fix it? I didn't have time to close issue about cffi. Hope will do during this week. It doesn't look urgent.

**Comment by mkudinov** [comment]
Yes, I missed several pieces of code with ctypes function calls so cffi crashed

Sure, when you have time
